Added the ability to filter events based on boolean expression(s) in the Events Window.

MokoSan opened the pull request:
## Summary

Added the ability to filter events in the event view using boolean expression(s). This PR Addresses a large part of https://github.com/microsoft/perfview/issues/1629 (CC: @Maoni0). In a nutshell, this PR allows the ability to specify a filter query (currently only) in the Columns To Display textbox to apply to all events so that we can easily filter the data to retrieve pertinent events.

## Design Details

In a nutshell, here's what's happening under the hood:

1. If the user specifies a filter query within ``[]`` in the "Columns To Display" textbox, we extract the intended query and create a ``FilterQueryExpressionTree`` out of it based on the string provided by the user.
    * A ``FilterQueryExpressionTree`` encapsulates the collection of ``FilterQueryExpression``s and the matching of the trace event on the collection of these expressions. 
    * ``FilterQueryExpression``s are of the following format: ``LeftOperand Operator RightOperand`` where:
        - ``LeftOperand`` can either be a name of a property to search on in any event e.g. ``ThreadID`` or be specified specifically for a particular event name (we are simply checking if the event name passed in by the user is contained in the full ``EventName`` property of the TraceEvent) by using ``::``. For example: ``CSwitch::OldThreadID``.
        - ``Operator`` can be any of the following: ``!=, =, <, <=, >=, >, Contains``.
        - ``RightOperand`` can either be a ``string`` or a numeric that's used as a ``double``.
    * Multiple ``FilterQuerieExpression``s can be combined using either ``||`` or ``&&``. 
2. The algorithm used to deduce the logical expression is a modification of the [Shunting Yard Algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm) that takes in an infix expression, converts it into it's postfix (aka Reverse Polish Notation (RPN)) equivalent.
    * The design decision to convert to the RPN equivalent is based on making the boolean expression reduction more performant when we iterate through all TraceEvents to discern which ones to display.
3. Once the RPN expression is obtained, and the user invokes a UI update, the entire RPN expression is applied to the TraceEvents and a single boolean is returned on the basis of which, we decide to display or discard the particular event. 

While working on this PR, the solution was prototyped [here](https://github.com/MokoSan/BooleanExpressionParser). A number of unit tests have also been added with this PR.

## Current Implementation Limitations

1. The way I have designed this, I have imposed a restriction of 26 Filter Query Expressions. This is implementation specific and I'll be happy to change this, if need be. The trade-off here was to keep the construction of final RPN expression quick and easy to reduce in the more perf demanding part of the code.
2. To construct an expression, you _have_ to add a space between the Left Operand, Operator and Right Operand. e.g. ``Depth<1`` will not work - instead you'll need to enter: ``Depth < 1``. This is an implementation detail, as well; to construct ``FilterQueryExpression``s, I am splitting on empty spaces to distinguish between the different parts of the expression.
4. Right now, only _one_ query will be used to filter the event i.e. you can't specify multiple ``[..]``s in the Columns To Display textbox.

## Demo

### Demo of Greater + Less Than or Equal to Operators:
https://user-images.githubusercontent.com/4951960/168762353-416a619c-0c2f-4b59-887c-213a7fe2f77f.mp4

### Demo of More Complex Operations With A Bunch of Nesting:
https://user-images.githubusercontent.com/4951960/168765726-f6644215-77e3-4f5a-a70f-a1359a82cdc0.mp4

Some of the queries I demo'd based on the ``GC/Start`` and ``GC/Stop`` events are:
1. ``[(Count > 10 && (Depth >= -1) && (Count <= 30) && (Count <= 30 || ProcessorNumber = 2))] Count ProcessorNumber Depth``
2. ``[(Count > 10  && (Count <= 30) && (Count <= 30 || ProcessorNumber = 2))] Count ProcessorNumber Depth``
5. ``[(Count > 10  && (Count <= 30) && (Count <= 30 || ProcessorNumber = 2))] Count ProcessorNumber``
6. ``[(Count > 10  && (Count <= 30)  && (Count <= 30 || ProcessorNumber = 2))] Count ProcessorNumber``

Happy to address any feedback!

brianrob commented:
@MokoSan, this is super cool.  I am playing around with this locally and starting to look at the code.

MokoSan commented:
> @MokoSan, this is super cool. I am playing around with this locally and starting to look at the code.

Thanks for taking a look! Please let me know if you have any feedback. I am also working on getting this functionality working from the Find text box; might submit that as a separate PR.

MokoSan commented:
Hi @brianrob, I think this PR is ready to be merged sans the seemingly unrelated build failures; have added the ability to do the query search for the Find text box and added a plethora of unit tests. I was unable to figure out how to rerun the CI checks. How can I do so? Thanks for your help!

brianrob commented on src/CSVReader/EventSource.cs:
```diff
@@ -90,16 +95,22 @@ public abstract class EventSource
+        /// all possible columns and returns the list that match the specification. This function also extracts the event filter query and creates a FilterQueryExpressionTree.
         /// </summary>
-        public static List<string> ParseColumns(string columnSpec, ICollection<string> columnNames)
+        public static List<string> ParseColumns(string columnSpec, ICollection<string> columnNames, out FilterQueryExpressionTree filterQueryExpressionTree)
```
This looks like a breaking change for any extension that might use the events view to show non-standard data.  Do you think that it is possible to remove the new parameter and depend on the `FilterQueryExpressionTree` property?  From my review so far, I'm not sure how to think about the property vs. this out parameter.

brianrob commented on src/PerfView/EtwEventSource.cs:
```diff
@@ -327,6 +327,15 @@ public override void ForEach(Func<EventRecord, bool> callback)
                         }
                     }
 
+                    if (FilterQueryExpressionTree != null)
```
I like that this is completely opt-in and so if a user doesn't specify a filter, then none of this runs.

brianrob commented on src/PerfView/EventViewer/EventWindow.xaml.cs:
```diff
@@ -1108,20 +1112,64 @@ public bool Find(string pat)
+                        }
+                    }
+
+                    foundItem = tree.Match(data, item.EventName);
```
I'm wondering if it makes sense to pass the entire `EventRecord` in here and look for ways to optimize here, such as not parsing `Rest` until we have to.

MokoSan replied:
>> I'm wondering if it makes sense to pass the entire EventRecord in here

Definitely thought about that. I held back so far because: I figured I leave the input data type to be PerfView construct agnostic; specifically, the ``EventRecord`` type exists in the PerfView project but since I am building this functionality in the TraceEvent project I would either have to move the entire implementation from TraceEvent to PerfView or move the ``EventRecord`` definition to the TraceEvent library. 

My thinking here was that if there could be other uses of this filtering mechanism, might as well have it exposed as behavior in the TraceEvent side of things such as for cases where we'd possibly want to programmatically filter events outside Perfview. 

I'd be curious to know your opinion here and could definitely move things around. 

>>  such as not parsing Rest until we have to.

Definitely, the latest implementation takes care of this where we are only parsing the Rest column if the rest of the checks fail.

brianrob commented on src/TraceEvent/TraceUtilities/FilterQueryExpression/FilterQueryExpressionTree.cs:
```diff
@@ -0,0 +1,87 @@
+            }
+
+            // Map each of the expressions to lower-case alphabet and the match result of the expression i.e. a boolean.
+            Dictionary<string, bool> convertedExpressionMap = new Dictionary<string, bool>();
```
Same here - wondering if we can cache this or otherwise re-use it.

brianrob reviewed (commented):
@MokoSan, thank you very much for building this.  It's great!  For the most part, I am pretty happy with how this is architected.  A few thoughts that I have as I reviewed the code and tried it out:

1. I like that this is fairly self-contained and doesn't run at all if you don't specify a filter.  That's great from a "if we mess something up we don't affect existing functionality" perspective.
2. I tried this out on my machine and it's super cool.  One thing I noticed is that the code is very whitespace particular.  For example, `[(ProcessorNumber = 2)]` works but `[(ProcessorNumber =2)]` throws an exception.  I also noticed that if you don't specify any fields then you get an empty `Rest` column.  Is that intentional?
3. I am a bit leary of perf.  I see lots of allocations of dictionaries, calls to `string.Split`, etc. and some N-squared algorithms (e.g. looking up each field in a set of field names, which results in a O(N) lookup per-field).  I am wondering how this behaves over a reasonably large trace.  I'm thinking that there is likely room for optimization, but I'm wondering if you've profiled any of this.  I want to be pragmatic and not force perfection here - just want to make sure that we aren't egregious.  :)

Happy to chat about these and work with you on plans going forward.  Thank you again for this - it's going to be a great addition.

MokoSan commented:
Hi @brianrob, thank you very much for taking the time to review the PR! 

Here are my responses:

> 2. I tried this out on my machine and it's super cool.  One thing I noticed is that the code is very whitespace particular.  For example, `[(ProcessorNumber = 2)]` works but `[(ProcessorNumber =2)]` throws an exception.  I also noticed that if you don't specify any fields then you get an empty `Rest` column.  Is that intentional?

Yes, this is intentional - I mentioned this in the first comment in the PR but let me know if I should be adding this anywhere else to better broadcast the usage. 

**Details**: The way I have this currently implemented is by splitting on empty spaces to deduce the operands and the operator. Do let me know if this should be changed. I have a couple of ideas but will require some more work.  

> 3. I am a bit leary of perf.  I see lots of allocations of dictionaries, calls to `string.Split`, etc. and some N-squared algorithms (e.g. looking up each field in a set of field names, which results in a O(N) lookup per-field).  I am wondering how this behaves over a reasonably large trace.  I'm thinking that there is likely room for optimization, but I'm wondering if you've profiled any of this.  I want to be pragmatic and not force perfection here - just want to make sure that we aren't egregious.  :)

Definitely will be happy to iterate on this further. I have addressed a majority of your comments with the latest update.

## Microbenchmarking 

I used Benchmark.NET to get details about the wall-clock time and allocations based on creating and matching simple and complex queries. 

NOTE: By clicking on the ▸in any of the sections below, you should have access to all the Benchmark.NET results and code.

### For Simple Queries 

For simple queries, the average wall-clock time it takes to create a filter query tree and to match on it by a simple query is ~1.5μs allocating ~1.2KB per call.

<details>

#### Results

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
11th Gen Intel Core i7-1165G7 2.80GHz, 1 CPU, 8 logical and 4 physical cores
  [Host]                                    : .NET Framework 4.8 (4.8.4510.0), X86 LegacyJIT  [AttachedDebugger]
  FilterQueryTreePerf_CreateAndMatch_Simple : .NET Framework 4.8 (4.8.4510.0), X86 LegacyJIT

Job=FilterQueryTreePerf_CreateAndMatch_Simple  InvocationCount=100  IterationCount=100  
UnrollFactor=1  

```
|                          Method |    SimpleExpression |       Mean |     Error |    StdDev |     Median |        Min |         Q1 |         Q3 |        Max | Allocated |
|-------------------------------- |-------------------- |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-----------:|-----------:|----------:|
| **CreateAndMatch_SimpleExpression** |    **ThreadID = 1,001** | **1,245.3 ns** | **105.53 ns** | **311.17 ns** | **1,136.0 ns** |   **830.0 ns** | **1,019.5 ns** | **1,464.0 ns** | **1,961.0 ns** |     **901 B** |
| **CreateAndMatch_SimpleExpression** |    **ThreadID &gt; 1,000** | **1,599.4 ns** | **128.19 ns** | **373.93 ns** | **1,547.5 ns** |   **974.0 ns** | **1,304.0 ns** | **1,889.8 ns** | **2,407.0 ns** |     **901 B** |
| **CreateAndMatch_SimpleExpression** |   **ThreadID &gt;= 1,001** | **1,428.0 ns** |  **82.64 ns** | **239.75 ns** | **1,447.5 ns** | **1,033.5 ns** | **1,207.5 ns** | **1,641.5 ns** | **1,921.5 ns** |     **983 B** |
| **CreateAndMatch_SimpleExpression** | **ThreadID Contains 1** |   **962.4 ns** |  **27.08 ns** |  **76.82 ns** |   **945.0 ns** |   **831.0 ns** |   **903.0 ns** | **1,013.0 ns** | **1,200.0 ns** |     **983 B** |

#### Code

```csharp
    [SimpleJob(targetCount: 100, invocationCount: 100, id: "FilterQueryTreePerf_CreateAndMatch_Simple")]
    [MinColumn, Q1Column, MedianColumn, Q3Column, MaxColumn]
    [MemoryDiagnoser]
    [MarkdownExporter]
    public class FilterQueryTreePerf_CreateAndMatch_Simple
    {
        [Params("ThreadID = 1,001", "ThreadID >= 1,001", "ThreadID > 1,000", "ThreadID Contains 1")]
        public string SimpleExpression { get; set; }

        [Benchmark]
        public void CreateAndMatch_SimpleExpression()
        {
            FilterQueryExpressionTree tree = new FilterQueryExpressionTree(SimpleExpression);
            var data = new Dictionary<string, string>()
            {
                { "ThreadID", "1,001" },
            };
            tree.Match(data, "GC/Start");
        }
    }
```

</details>

### For Complex Queries

For more complex queries, the average wall-clock time varies fairly significantly depending on the query. E.g. for a fairly nested complex query such as: ``((GC/Start::Depth = 10 && (GC/Start::Depth <= 11 && (GC/Start::Depth <= 12 && (Depth <= 13 && (GC/Start::Depth <= 14))))))`` , we average around 12.5μs of wall-clock time and allocate 9KB of memory as opposed to a query like: ```"((Depth <= 10) && (Depth <= 20)) || ((Depth != 15) && (Depth > 5))"`` where we average around 7.22μs and allocate 6KB. 

<details>

#### Results

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
11th Gen Intel Core i7-1165G7 2.80GHz, 1 CPU, 8 logical and 4 physical cores
  [Host]                            : .NET Framework 4.8 (4.8.4510.0), X86 LegacyJIT  [AttachedDebugger]
  FilterQueryTreePerf_Match_Complex : .NET Framework 4.8 (4.8.4510.0), X86 LegacyJIT

Job=FilterQueryTreePerf_Match_Complex  InvocationCount=100  IterationCount=100  
UnrollFactor=1  

```
|                           Method |    ComplexExpression |      Mean |     Error |    StdDev |    Median |      Min |        Q1 |        Q3 |       Max | Allocated |
|--------------------------------- |--------------------- |----------:|----------:|----------:|----------:|---------:|----------:|----------:|----------:|----------:|
| **CreateAndMatch_ComplexExpression** | **((Dep(...)&gt; 5)) [66]** |  **7.224 μs** | **0.1879 μs** | **0.5269 μs** |  **7.043 μs** | **6.484 μs** |  **6.893 μs** |  **7.424 μs** |  **8.817 μs** |      **6 KB** |
| **CreateAndMatch_ComplexExpression** |  **((GC(...))))) [122]** | **12.576 μs** | **0.8231 μs** | **2.3350 μs** | **11.783 μs** | **9.009 μs** | **10.781 μs** | **13.876 μs** | **20.172 μs** |      **9 KB** |
| **CreateAndMatch_ComplexExpression** | **((GC/(...) 10)) [92]** |  **8.980 μs** | **0.4311 μs** | **1.2299 μs** |  **8.584 μs** | **7.514 μs** |  **8.077 μs** |  **9.632 μs** | **12.423 μs** |      **7 KB** |
| **CreateAndMatch_ComplexExpression** | **((Thr(...)1001) [68]** |  **5.686 μs** | **0.2163 μs** | **0.6171 μs** |  **5.372 μs** | **5.008 μs** |  **5.232 μs** |  **5.922 μs** |  **7.278 μs** |      **5 KB** |
| **CreateAndMatch_ComplexExpression** | **((Thr(...)1001) [45]** |  **4.266 μs** | **0.3320 μs** | **0.9631 μs** |  **3.907 μs** | **3.209 μs** |  **3.526 μs** |  **4.858 μs** |  **7.092 μs** |      **3 KB** |
| **CreateAndMatch_ComplexExpression** | **(Dept(...)= 20) [30]** |  **3.746 μs** | **0.1663 μs** | **0.4744 μs** |  **3.582 μs** | **3.216 μs** |  **3.379 μs** |  **3.997 μs** |  **5.059 μs** |      **3 KB** |
| **CreateAndMatch_ComplexExpression** | **(GC/S(...)s 10) [50]** |  **4.714 μs** | **0.2925 μs** | **0.8202 μs** |  **4.503 μs** | **3.763 μs** |  **4.039 μs** |  **5.058 μs** |  **7.478 μs** |      **4 KB** |
| **CreateAndMatch_ComplexExpression** | **(GC/S(...)= 20) [48]** |  **5.564 μs** | **0.4654 μs** | **1.3577 μs** |  **5.129 μs** | **3.916 μs** |  **4.441 μs** |  **6.520 μs** |  **9.515 μs** |      **4 KB** |
| **CreateAndMatch_ComplexExpression** | **(OldP(...)est2) [77]** |  **6.068 μs** | **0.2337 μs** | **0.6704 μs** |  **5.792 μs** | **5.370 μs** |  **5.574 μs** |  **6.674 μs** |  **8.434 μs** |      **6 KB** |
| **CreateAndMatch_ComplexExpression** | **Depth(...)!= 10 [41]** |  **4.936 μs** | **0.3329 μs** | **0.9552 μs** |  **4.509 μs** | **3.970 μs** |  **4.222 μs** |  **5.545 μs** |  **7.585 μs** |      **4 KB** |
| CreateAndMatch_ComplexExpression | Depth(...)!= 10 [41] |  5.365 μs | 0.3979 μs | 1.1289 μs |  4.941 μs | 3.942 μs |  4.538 μs |  5.749 μs |  8.388 μs |      4 KB |
| **CreateAndMatch_ComplexExpression** | **Depth(...)&lt;= 20 [26]** |  **3.672 μs** | **0.2279 μs** | **0.6576 μs** |  **3.416 μs** | **2.957 μs** |  **3.165 μs** |  **4.032 μs** |  **5.578 μs** |      **3 KB** |

#### Code

```csharp 
    [SimpleJob(targetCount: 100, invocationCount: 100, id: "FilterQueryTreePerf_Match_Complex")]
    [MinColumn, Q1Column, MedianColumn, Q3Column, MaxColumn]
    [MemoryDiagnoser]
    [MarkdownExporter]
    public class FilterQueryTreePerf_CreateAndMatch_Complex
    {
        [Params("(Depth >= 10) && (Depth <= 20)",
        "Depth <= 10 || Depth <= 20",
        "((Depth <= 10) && (Depth <= 20)) || ((Depth != 15) && (Depth > 5))",
        "Depth != 10 && Depth != 10 && Depth != 10",
        "Depth <= 10 && Depth <= 10 && Depth != 10",
        "((ThreadData != 1,0033) && ThreadData = 1001)",
        "((ThreadData != 1,0033 && ThreadData != 1,002) && ThreadData = 1001)",
        "(OldProcessName = test && OldProcessName != test1 || OldProcessName != test2)",
        "(GC/Start::Depth >= 10 && GC/Start::Depth <= 20)",
        "((GC/Start::Depth = 10 && (GC/Start::Depth <= 11 && (GC/Start::Depth <= 12 && (Depth <= 13 && (GC/Start::Depth <= 14))))))",
        "((GC/Start::Depth = 10) && (Depth <= 10)) || ((GC/Start::ThreadData != 10) || (Depth >= 10))",
        "(GC/Start::Depth = 10) && (ThreadData Contains 10)")]
        public string ComplexExpression { get; set; }

        [Benchmark]
        public void CreateAndMatch_ComplexExpression()
        {
            FilterQueryExpressionTree tree = new FilterQueryExpressionTree(ComplexExpression);
            var data = new Dictionary<string, string>
                {
                    { "Depth", "10" },
                    { "ThreadData", "10" }
                };
            tree.Match(data, "GC/Start");
        }
    }
```

</details>

## CPU Analysis Based On Tracing PerfView

I collected a trace that examined another trace with the events window open with ~1M+ CSwitch events with the following Columns to display textbox filled:

Cols: ``OldProcessName ProcessorNumber [(ProcessName = devenv) && (ProcessorNumber = 5)]``

<img width="960" alt="image" src="https://user-images.githubusercontent.com/4951960/173233637-b6e9b8be-f756-4fd1-81a3-1188aea0aed7.png">
<img width="365" alt="image" src="https://user-images.githubusercontent.com/4951960/173233645-8f52a4c4-4b6b-47b1-bdcd-7b458ad1211e.png">

The results indicated that to traverse 1M events and to filter 10k out of them for a 2 expression query, it takes ~1 second of CPU time to Match (this was much lesser than what I expected). Additionally, for the Shunting Yard Algorithm, the lifeblood of the boolean expression deduction, ``string[] rawTokens = postfixNotation.Split(' ');`` and `` if (_operandRegex.IsMatch(rawToken))`` were the two most expensive calls. For Matching individual queries, Splitting the query expression and looking in the payload of the Trace Events took the most time.

I wouldn't call these egregious with these latest changes but do let me know if you'd want me to optimize further.

Further details:

Name | Inc % | Inc | Exc % | Exc
-- | -- | -- | -- | --
microsoft.diagnostics.tracing.traceevent!FilterQueryExpressionTree.Match | 28.3 | 1,817 | 0.3 | 18
| +   microsoft.diagnostics.tracing.traceevent!ShuntingYard.Match | 14.3 | 919 | 0.2 | 10
|+   microsoft.diagnostics.tracing.traceevent!FilterQueryExpression.Match | 11.2 | 721 | 0.3 | 17
|+   mscorlib.ni!System.String.CtorCharCount(Char, Int32) | 1.1 | 70 | 0.0 | 0
|+   mscorlib.ni!System.Collections.Generic.Dictionary`2[System.__Canon,System.Boolean].Insert(System.__Canon,   Boolean, Boolean) | 0.6 | 39 | 0.5 | 31
|+   mscorlib.ni!System.Collections.Generic.Dictionary`2[System.__Canon,System.Boolean].Clear() | 0.5 | 34 | 0.0 | 3
|+   mscorlib.ni!System.Collections.Generic.Dictionary`2+Enumerator[System.Char,System.__Canon].MoveNext() | 0.2 | 13 | 0.2 | 12

Zooming in on the Shunting Yard algorithm, the lifeblood of the matching side of things:

Name | Inc % | Inc | Exc % | Exc
-- | -- | -- | -- | --
| +   microsoft.diagnostics.tracing.traceevent!ShuntingYard.Match | 14.3 | 919 | 0.2 | 10
|+   system.ni!Regex.IsMatch | 6.6 | 422 | 0.0 | 0
|+   mscorlib.ni!String.SplitInternal | 3.8 | 247 | 0.0 | 2

Zooming in on the individual query Match side of things:

Name | Inc % | Inc | Exc % | Exc
-- | -- | -- | -- | --
|+   microsoft.diagnostics.tracing.traceevent!FilterQueryExpression.Match | 11.2 | 721 | 0.3 | 17
|+ mscorlib.ni!String.Split | 5.5 | 351 | 0.1 | 9
|+   microsoft.diagnostics.tracing.traceevent!FilterQueryUtilities.ExtractPayloadByName | 5.4 | 349 | 0.1 | 9
|+ mscorlib.ni!String.EqualsHelper | 0.0 | 2 | 0.0 | 2

The command I used to collect the trace: ``.\PerfView.exe /nogui /accepteula collect withperf.etl``.

MokoSan commented:
Hi @brianrob,

I believe I have addressed the following issues in the latest commit:

- Added documentation in the ``Columns To Display`` and ``Find`` sections.
- Removed the requirement of providing spaces amongst the operands and the operator per expression i.e. we can now write expressions such as ``[(ProcessName=devenv) && (GC/Start::Depth >=2)]``. 
- Added more robustness to all the Filter Query Expression calls by wrapping around try-catch blocks and writing to the log file in case of an exception rather than bubbling it up to the user.

Do let me know if you have any further feedback! Thanks for your help!

brianrob commented on src/PerfView/EventViewer/EventWindow.xaml.cs:
```diff
@@ -1075,7 +1075,21 @@ public bool Find(string pat)
+            FilterQueryExpressionTree tree;
+            try
+            {
+                string modifiedPat = FilterQueryUtilities.TryExtractFilterQueryExpression(pat, out tree);
```
I did some additional testing and found an example that didn't work.  If you filter the view to just AllocationTick events, and then use the following expression `[TypeName = System.Char[]]`, I get no results even if they do exist.  I can confirm this with `[TypeName contains System.Char]`.  I'm assuming this is because the type is an array.

MokoSan replied:
Yes, I think this is expected behavior with the current implementation. Unfortunately, I have developed this in a way where we are matching only on the first [] and ignoring the rest. I have a couple of ideas as to how this can be fixed (I'll take this on post-merge of this PR if that's fine with you).

brianrob commented on src/PerfView/EventViewer/EventWindow.xaml.cs:
```diff
@@ -1286,6 +1300,11 @@ public void Update()
             }
+            catch(Exception ex)
+            {
+                StatusBar.LogError(ex.Message);
```
I don't think this quite did what I wanted - the error seems to be overwritten by the number of records found, but it does at least write something to the log, so that's good.  I think this will be sufficient for now, and we can see if we need to something here in the future.

brianrob reviewed (commented):
@MokoSan, this is looking really good.  A couple of comments - I think most of these should be reasonably easy, but let me know if you want to defer any of these to post merge.

brianrob reviewed (approved):
LGTM!
